pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -270,19 +509,46 @@ void Stencil::DoStenciling() {
   CollectUsedIndices();
   CollectStrideOneIndices();
 
+  auto ranges = op.getConstantRanges();
+  if (!ranges)
+    return;
+
+  opConstRanges = *ranges;
+
+  // Search tensors' order, inner index and their tiles
+  SearchTensorsOrder();
+
+  IVLOG(1, "Best Perf: " << bestPerf);
+  IVLOG(1, "Best Tiles: " << bestTiles[0] << ":" << bestTiles[1] << ":"
+                          << bestTiles[2]);
+
+  if (bestPerf == std::numeric_limits<double>::max()) {
+    IVLOG(1, "No tile plan for stencil.");
+    return;
+  }
+
   op.setAttr("is_gemm", mlir::UnitAttr::get(op.getContext()));
 }
 
 void StencilPass::runOnFunction() {
   auto func = getFunction();
   func.walk([&](mlir::AffineParallelOp op) {
-    Stencil as(op);
+    unsigned threads = numThreads.getValue();
+    if (threads == 0) {
+      threads = std::thread::hardware_concurrency();
+    }
+    Stencil as(op, threads);
     as.DoStenciling();
   });
 }
 
-std::unique_ptr<mlir::Pass> createStencilPass() {
-  return std::make_unique<StencilPass>();
+StencilPass::StencilPass(const StencilPassOptions &options) {
+  numThreads.setValue(options.numThreads);
+}
+
+void createStencilPass(mlir::OpPassManager &pm,
+                       const StencilPassOptions &options) {
```

Comment:
I don't think we need a separate options object, can we just pass `numThreads` here as an argument?

Reply:
I do foresee adding more options here. Passing the generic Options object allows us to pull all the things will need from it. Besides, the call to create the pass (creating a pass with instance Options is much different than the one with no options) is created like the following (much different than the pass with no options - for the options to be parsed and processed the pass is added to the OpPassManager)...
pm.addPass(std::make_unique<StencilPass>(options));
In no option pass, the pass is created only by doing std::make_unique<PadPass>();
The idiomatic boilerplate code for creating and instantiating a pass with instance specific options comes from the MLIR doc - https://github.com/llvm/llvm-project/blob/master/mlir/docs/WritingAPass.md#instance-specific-pass-options

Reply:
So I think we want to register a pass instead of a pipeline here. The first example in that doc is for passes, the second example is for pipelines.

Reply:
Yes, there are two different examples in that section of the docs, passes and pass pipelines, only the first one is relevant for us.  Certainly the wording could have been clearer in the docs, or perhaps a section break would be helpful, I misread that one the first time too...

Reply:
The checkin before the last one doesn't use the PipelinePass anymore.